src/chainstate/stacks/index/trie_sql.rs

```diff
@@ -0,0 +1,293 @@
+    w.write_all(&hash_buff)
+        .map_err(|e| e.into())
+}
+
+pub fn read_node_type(conn: &Connection, block_id: u32, ptr: &TriePtr) -> Result<(TrieNodeType, TrieHash), Error> {
+    let mut blob = conn.blob_open(rusqlite::DatabaseName::Main, "marf_data", "data", block_id.into(), true)?;
+    read_nodetype(&mut blob, ptr)
+}
+
+pub fn get_node_hash_bytes(conn: &Connection, block_id: u32, ptr: &TriePtr) -> Result<TrieHash, Error> {
+    let mut blob = conn.blob_open(rusqlite::DatabaseName::Main, "marf_data", "data", block_id.into(), true)?;
+    let hash_buff = bits_read_node_hash_bytes(&mut blob, ptr)?;
+    Ok(TrieHash(hash_buff))
+}
+
+pub fn get_node_hash_bytes_by_bhh(conn: &Connection, bhh: &BlockHeaderHash, ptr: &TriePtr) -> Result<TrieHash, Error> {
+    let row_id: i64 = conn.query_row("SELECT block_id FROM marf_data WHERE block_hash = ?",
+                                     &[bhh], |r| r.get("block_id"))?;
+    let mut blob = conn.blob_open(rusqlite::DatabaseName::Main, "marf_data", "data", row_id, true)?;
+    let hash_buff = bits_read_node_hash_bytes(&mut blob, ptr)?;
+    Ok(TrieHash(hash_buff))
+}
+
+pub fn get_chain_tips(conn: &Connection) -> Result<Vec<BlockHeaderHash>, Error> {
+    let mut s = conn.prepare("SELECT block_hash FROM chain_tips")?;
+    let rows = s.query_map(NO_PARAMS, |row| row.get("block_hash"))?;
+    rows.map(|i| i.map_err(|e| e.into())).collect()
+}
+
+pub fn remove_chain_tip_if_present(conn: &Connection, bhh: &BlockHeaderHash) -> Result<(), Error> {
+    conn.execute("DELETE FROM chain_tips WHERE block_hash = ?", &[bhh])?;
+    Ok(())
+}
+
+pub fn add_chain_tip(conn: &Connection, bhh: &BlockHeaderHash) -> Result<(), Error> {
+    conn.execute("INSERT INTO chain_tips (block_hash) VALUES (?)", &[bhh])?;
+    Ok(())
+}
+
+pub fn retarget_chain_tip_entry(conn: &Connection, from: &BlockHeaderHash, to: &BlockHeaderHash) -> Result<(), Error> {
+    conn.execute("UPDATE chain_tips SET block_hash = ? WHERE block_hash = ?",
+                 &[to, from])?;
+    Ok(())
+}
+
+pub fn lock_bhh_for_extension(conn: &mut Connection, bhh: &BlockHeaderHash) -> Result<bool, Error> {
+    let tx = conn.transaction()?;
+    let is_bhh_committed = tx.query_row("SELECT 1 FROM marf_data WHERE block_hash = ? LIMIT 1", &[bhh],
+                                        |_row| ()).optional()?.is_some();
+    if is_bhh_committed {
+        return Ok(false)
+    }
+
+    let is_bhh_locked = tx.query_row("SELECT 1 FROM block_extension_locks WHERE block_hash = ? LIMIT 1", &[bhh],
+                                     |_row| ()).optional()?.is_some();
+    if is_bhh_locked {
+        return Ok(false)
+    }
+
+    tx.execute("INSERT INTO block_extension_locks (block_hash) VALUES (?)", &[bhh])?;
+
+    tx.commit()?;
+    Ok(true)
+}
+
+pub fn count_blocks(conn: &Connection) -> Result<u32, Error> {
+    let result = conn.query_row("SELECT COUNT(1) AS count FROM marf_data", NO_PARAMS, |row| row.get("count"))?;
```

Comment:
Last I checked, `SELECT COUNT(..)` in sqlite requires a full DB scan.  If we're just trying to get a block count, why not set `block_id` to `PRIMARY KEY AUTOINCREMENT` and rewrite this query as `SELECT MAX(block_id)`?  We never delete blocks, so `MAX(block_id)` should always be the number of blocks for the foreseeable future.

Reply:
Yep, that's true. I always want SELECT COUNT to be smarter about tables with integer primary keys, but they're not. I will update -- MAX(block_id)+1 should would here.

Reply:
Updated in c0fb0418d10800e764f88694b24e12ef917cdf39